students/jjakub/session05/mailroom03.py

```diff
@@ -0,0 +1,84 @@
+
+
+import sys
+import tempfile
+
+user_options = {}
+user_options = {
+"Send": "Send a Thank You to a single donor" ,
+"Report": "Create a Report" ,
+"Letter": "Send a letter to all donors" ,
+"Quit": "Quit"
+}
+
+donor_db  = {}
+donor_db = {
+"alex" : [53898.23, 2653.00, 105.07] ,
+"jeff" : [183.00, 200.00] ,
+"paul" : [33941.98] ,
+"mark" : [1.65, 55.00] ,
+"john" : [2705.71]}
+
+def sort_key(donor_db):
+    return sum(donor_db[1])
+
+def send_thanks():
+    name = ''
+    while not name or name.lower() == 'list':
+        name = input("Please enter the full name: ").lower()
+        if name.lower() == 'list':
+            print(' '.join(donor_db))
+    while True:
+        try:
+            amt = int(input("Please enter the donation amount: "))
+            donor_db.setdefault(name,[]).append(amt)
+            print(f'\nDear {name.capitalize()},\nThank you for the donation of ${amt:.2f}.\nSincerely, \nThe Mailroom Foundation')\
+            return
+        except ValueError:
+            print("\nWarning: Donation amount must be an integer")
+
+def donor_rpt():
+    header = ['Donor Name', 'Total Given', 'Num Gifts', 'Average Gift']
+    seperator = '-' * 50
+    print(f'\n{header[0]:13}|{header[1]:<10}|{header[2]:<10}|{header[3]:<10}\n{seperator}')
+    donor_db_sum = {donor: sum(donor_db[donor]) for donor in donor_db}
+    key_sort = sorted(donor_db_sum, key=donor_db_sum.__getitem__,reverse=True)
+    for donor in key_sort:
+        print(f'{donor.capitalize():13}', f'{sum(donor_db.get(donor)):>10.2f}', f'{len(donor_db.get(donor)):>10}', f'${sum(donor_db.get(donor))/len(donor_db.get(donor)):>10.2f}')
+
+def thanks_all():
+    path = input("Please enter the path: ")
+    if path != '':
+        path = path + "\\"
+    for donor_key,donor_val in donor_db.items():
+        with open("{}{}.txt".format(path, donor_key), "w") as thank_you:
+            thank_you.write(f'\nDear {donor_key.capitalize()},\nThank you for the donation of ${sum(donor_db.get(donor_key)):.2f}.\nSincerely, \nThe Mailroom Foundation')
```

Comment:
A good place to use `try`-`except` for cases when the file writing fails (e.g., due to `PermissionError` or `IOError`).